src/bresult.ml

```diff
@@ -1,194 +1,233 @@
-  | Ok _ as res -> res
-  | Error y -> Error (f y)
+let guardToString f =
+  try Ok (f())
+  with e -> fromException e
 
-let map2 f g e = match e with
-  | Ok x -> Ok (f x)
-  | Error s -> Error (g s)
+let guardToStringTrace f =
+  try Ok (f())
+  with e -> fromExceptionTrace e
 
-let iter f e = match e with
-  | Ok x -> f x
-  | Error _ -> ()
+let wrap f x =
+  try make (f x)
+  with e -> Error e
 
-exception Get_error
+let wrap2 f x y =
+  try make (f x y)
+  with e -> Error e
 
-let get_exn = function
-  | Ok x -> x
-  | Error _ -> raise Get_error
+let wrap3 f x y z =
+  try make (f x y z)
+  with e -> Error e
 
-let get_or e ~default = match e with
+
+let isOk = function
+  | Ok _ -> true
+  | Error _ -> false
+
+let isError = function
+  | Ok _ -> false
+  | Error _ -> true
+
+let equal ?(err=Pervasives.(=)) eq a b = match a, b with
+  | Ok x, Ok y -> eq x y
+  | Error e, Error e' -> err e e'
+  | _ -> false
+
+let compare ?(err=Comparison.compare) cmp a b = match a, b with
+  | Ok x, Ok y -> cmp x y
+  | Ok _, _  -> Comparison.Greater
+  | _, Ok _ -> Comparison.Less
+  | Error e, Error e' -> err e e'
+
+
+exception GetError
+
+let get default = function
   | Ok x -> x
   | Error _ -> default
 
-let map_or f e ~default = match e with
-  | Ok x -> f x
+let getOrRaise = function
+  | Ok x -> x
+  | Error _ -> raise GetError
+
+let getOr ~default = function
+  | Ok x -> x
   | Error _ -> default
 
-let catch e ~ok ~err = match e with
-  | Ok x -> ok x
-  | Error y -> err y
+let getLazy defaultFn = function
+  | Ok x -> x
+  | Error _ -> defaultFn ()
 
-let flat_map f e = match e with
+
+let forEach f e = match e with
   | Ok x -> f x
-  | Error s -> Error s
+  | Error _ -> ()
 
-let (>|=) e f = map f e
+let map f = function
+  | Ok x -> Ok (f x)
+  | Error e -> Error e
 
-let (>>=) e f = flat_map f e
+let mapOr ~default f = function
+  | Ok x -> f x
+  | Error _ -> default
 
-let equal ?(err=Pervasives.(=)) eq a b = match a, b with
-  | Ok x, Ok y -> eq x y
-  | Error s, Error s' -> err s s'
-  | _ -> false
+let mapOrLazy ~default f = function
+  | Ok x -> f x
+  | Error _ -> default ()
 
-let compare ?(err=Pervasives.compare) cmp a b = match a, b with
-  | Ok x, Ok y -> cmp x y
-  | Ok _, _  -> 1
-  | _, Ok _ -> -1
-  | Error s, Error s' -> err s s'
+let mapError f = function
+  | Ok _ as res -> res
+  | Error e -> Error (f e)
 
-let fold ~ok ~error x = match x with
-  | Ok x -> ok x
-  | Error s -> error s
+let maybe f default = mapOr ~default f
 
-let is_ok = function
-  | Ok _ -> true
-  | Error _ -> false
+let map2 f a b = match a, b with
+  | Error e, _
+  | _, Error e -> Error e
+  | Ok x, Ok y -> Ok (f x y)
 
-let is_error = function
-  | Ok _ -> false
-  | Error _ -> true
+let mapEither f g = function
+  | Ok x -> Ok (f x)
+  | Error e -> Error (g e)
 
-(** {2 Wrappers} *)
+let catch ~ok ~err = function
+  | Ok x -> ok x
+  | Error e -> err e
 
-let guard f =
-  try Ok (f ())
-  with e -> Error e
+let flatMap f = function
```

Comment:
And it's named `flatMap` because it "flattens" a nested data structure. For a single-element container this just means not wrapping it, but generalized to data structures such as list, "flattening" eould involve merging lists.